### PR TITLE
Check changes are pushed before export nitrate

### PR DESCRIPTION
**Pull request by lukaszachy**
There is option to ignore this check, but user has been warned.

This check needs to be ignored in the testsuite.

Fixes: https://github.com/teemtee/tmt/issues/1272

@thrix You did similar 'validate_fmf_id' however it cannot be used as it will pass if object exist. But I'd like to check that all changes are committed and pushed as well. IMHO both methods can live together as they check slightly different things.

All: sorry for noise. THIS is the correct PR (I screwed up git push in the previous one)

### CODE REVIEW

**Comment by lukaszachy** [comment]
Fixed 'git ci' -> 'git commit' and rebased to main

**Comment by lukaszachy** [comment]
Fixed missing 'user.email' for git commit (so make rpm failed on %check)

**Comment by lukaszachy** [comment]
So even "user.name" is required :/ (+ rebased do the current main as well)

**Comment by lukaszachy** [comment]
Known TODO

- [ ] Check that .fmf/version marking used fmf root is committed
- [ ] test_all_good is fake, create a proper one
- [ ] add test_untracked_fmfroot
- [ ] add test_untracked_sources

File: tmt/utils.py [hunk 1]
```diff
@@ -1617,6 +1617,91 @@ def parse_yaml(content: str) -> EnvironmentType:
...
+
+    remote_ref = result.stdout.strip()
+
+    cmd = [
+        'git',
+        'diff',
+        f'HEAD..{remote_ref}',
+        '--name-status',
+        '--'] + sources
+    result = subprocess.run(
```
[the remark this answers is not in the document]
Comment (psss) [reply tmt/utils.py:1682]: Or, perhaps, we could just use `run = Common().run` here?

File: tests/unit/test_utils.py [hunk 2]
```diff
@@ -438,3 +440,123 @@ def test_FedoraDistGit(tmpdir):
...
     assert [("https://src.fedoraproject.org/repo/pkgs/rpms/tmt/fn-1.tar.gz/sha512/09af/fn-1.tar.gz",
             "fn-1.tar.gz")] == fedora_sources_obj.url_and_name(cwd=tmpdir)
+
+
+class Test_validate_git_status:
+    def test_all_good(cls, tmpdir):
+        pass
+
+    def test_no_remote(cls, tmpdir):
+        subprocess.run('git init'.split(), check=True, cwd=tmpdir)
```
Comment (psss) [inline tests/unit/test_utils.py:450]: Here also a `run()` function would make the code more readable. See also my comment in `tmt.utils`.

File: tests/unit/test_utils.py [hunk 3]
```diff
@@ -438,3 +440,123 @@ def test_FedoraDistGit(tmpdir):
...
+        subprocess.run('git init'.split(), check=True, cwd=origin)
+        subprocess.run(
+            'git config --local user.email lzachar@redhat.com'.split(),
+            check=True,
+            cwd=origin)
+        subprocess.run(
+            'git config --local user.name LZachar'.split(),
+            check=True,
+            cwd=origin)
+        fmfroot.join('main.fmf').write('test: echo')
```
Comment (psss) [inline tests/unit/test_utils.py:528]: What about creating a `setup()` method to cover these common preparation steps?

File: tests/integration/test_nitrate.py [hunk 4]
```diff
@@ -148,6 +148,36 @@ def test_missing_user_dryrun(self):
...
+
+        runner = CliRunner()
+
+        self.runner_output = runner.invoke(
+            tmt.cli.main,
+            ["test", "export", "--nitrate", "--debug", "--force-git-validation", "."],
+            catch_exceptions=False)
+
+        self.assertIn(
+            "Exporting regardless Not committed changes",
```
Comment (psss) [inline tests/integration/test_nitrate.py:178]:
```suggestion
            "Exporting regardless uncommitted changes",
```

The one-word form looks just a bit more straightforward?

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for improving this! Looks good in general. I've run into one possibly noop operation which could be just removed and propose to deduplicate some repeated stuff. Otherwise good.

File: tmt/cli.py [hunk 5]
```diff
@@ -552,6 +552,10 @@ def import_(
     '--bugzilla', is_flag=True,
     help="Link Nitrate case to Bugzilla specified in the 'link' attribute "
          "with the relation 'verifies'.")
+@click.option(
+    '--force-git-validation', is_flag=True,
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/cli.py:556]: Well, I want validation to happen each time. It has to be absolutely clear for user that nitrate -> tmt will not work as they intended.

File: tmt/utils.py [hunk 6]
```diff
@@ -1617,6 +1617,91 @@ def parse_yaml(content: str) -> EnvironmentType:
...
+    - without not committed local changes
+    - up to date on remote repository
+
+    When both check pass returns (True, '').
+    """
+    fmf_id = test.fmf_id
+    sources = test.node.sources
+    if 'path' in fmf_id:
+        # prepend fmf_id path but without leading '/'
+        sources = [os.path.join(fmf_id['path'][1:], s) for s in sources]
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/utils.py:1636]: Nice, thanks for finding this.

File: tests/unit/test_utils.py (hunk 3 above)
Comment (lukaszachy) [reply tests/unit/test_utils.py:528]: yep, fixture was created.

File: tmt/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/utils.py:1682]: `run = Common().run` works well, for some reason I though it needs to create workdir

File: tmt/cli.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (psss) [reply tmt/cli.py:556]: Perhaps a misunderstanding: I didn't mean to disable the check by default, it makes a perfect sense to do the check every time unless explicitly disabled if you really know what you're doing. Just that the option name (force git validation) does not fit the option description (ignore unpushed changes == skip git validation).

[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/cli.py:556]: How about --ignore-git-validation?

[the remark this answers is not in the document]
Comment (psss) [reply tmt/cli.py:556]: Yeap, that sounds much more aligned with the description.

**Comment by lukaszachy** [comment]
Renamed option, rebased to the current main + fix flake8 lines.

**Review by psss — APPROVED** [review]
Thanks for all the improvements! Looks good. Squashed.

**Pull request merged**